cmake/DefaultTargetProperties.cmake

```diff
@@ -0,0 +1,129 @@
+
+
+    #-------------------------------------------------------------------
+    # Symbol visibility
+
+    # Hide symbols from this library by default.
+    set_target_properties(
+        ${target_name}
+        PROPERTIES
+        C_VISIBILITY_PRESET hidden
+        CXX_VISIBILITY_PRESET hidden
+        VISIBILITY_INLINES_HIDDEN YES
+    )
+
+    # Hide all symbols from external statically linked libraries.
+    if (CMAKE_CXX_COMPILER_ID MATCHES ".*Clang" OR CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+        target_link_options(${target_name} PRIVATE -Wl,--exclude-libs,ALL)
+    endif ()
+
+    # Use C++11 ABI by default, but allow disabling.
+    if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND
+        CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 5.0)
+        option(GLIBCXX_USE_CXX11_ABI "For gcc, use the new C++11 library ABI" ON)
+        if (GLIBCXX_USE_CXX11_ABI)
+            target_compile_definitions(${target_name} PRIVATE _GLIBCXX_USE_CXX11_ABI=1)
+        else ()
+            target_compile_definitions(${target_name} PRIVATE _GLIBCXX_USE_CXX11_ABI=0)
+        endif ()
+    endif ()
+
+    # TODO(DF): POSITION_INDEPENDENT_CODE required for Python module? See OCIO.
+
+    #-------------------------------------------------------------------
+    # RPATH wrangling.
+
+    # With the 'usual' install path structure in mind, search from the
+    # bin directory (i.e. a binary loading a dynamic library) and then
+    # from the current directory (i.e. dynamic library loading another
+    # dynamic library).
+    if (APPLE)
+        set(rpath "@loader_path/../${CMAKE_INSTALL_LIBDIR};@loader_path")
+    else ()
+        set(rpath "$ORIGIN/../${CMAKE_INSTALL_LIBDIR};$ORIGIN")
+    endif ()
+
+    set_target_properties(
+        ${target_name}
+        PROPERTIES
+        # Allow CMake to set a temporary runtime library search path in
+        # the build phase that potentially hardcodes absolute paths,
+        # useful for TDD.
+        SKIP_BUILD_RPATH FALSE
+        BUILD_WITH_INSTALL_RPATH FALSE
+        # Don't add hardcoded directories to final runtime search path.
+        INSTALL_RPATH_USE_LINK_PATH FALSE
+        # Runtime search path value
+        INSTALL_RPATH "${rpath}"
+        # Enable RPATH on OSX
+        MACOSX_RPATH ON
+    )
+
+    #-------------------------------------------------------------------
+    # C++ #define to toggle static vs dynamic lib
+
+    # Used by header generated by CMake's `generate_export_header`
+    if (NOT BUILD_SHARED_LIBS)
+        target_compile_definitions(${target_name} PRIVATE ${PROJECT_NAME}_STATIC_DEFINE=1)
+    endif ()
+
+
+    #-------------------------------------------------------------------
+    # Build tree directory structure.
+
+    set_target_properties(
+        ${target_name}
+        PROPERTIES
+        # Make the build area layout look a bit more like the final dist
+        # layout
+        LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/${CMAKE_INSTALL_LIBDIR}
```

Comment:
Be cautious of using absolute paths here. Check the generated config files to ensure they do not contain absolute paths to the machine they are built on.

Config files should contain paths relative to a computed install prefix instead, which normally comes about in CMake by using relative paths in expressions such as this.

Reply:
You may be right. Worth checking. I've been bitten by having non-relocatable CMake config files generated too many times.

Reply:
The generated `OpenAssetIOConfig.cmake` has
```
get_filename_component(PACKAGE_PREFIX_DIR "${CMAKE_CURRENT_LIST_DIR}/../../../" ABSOLUTE)
```
then all other paths are relative to `PACKAGE_PREFIX_DIR`

Similarly `OpenAssetIOTargets.cmake` has
```
get_filename_component(_IMPORT_PREFIX "${CMAKE_CURRENT_LIST_FILE}" PATH)
```
and 
```
get_filename_component(_DIR "${CMAKE_CURRENT_LIST_FILE}" PATH)
```
and all other paths are relative to those.

So I think that means it is sufficiently relocatable.

Reply:
Agreed for the prefix, but ensure you also check the target_include_directories and target_link_libraries, as they can be set independently of the prefix depending on the CMake scripts. (Had to deal with this just this week.)